source/common/router/router.cc

```diff
@@ -848,28 +882,42 @@ Http::FilterDataStatus Filter::decodeData(Buffer::Instance& data, bool end_strea
   bool buffering = (retry_state_ && retry_state_->enabled()) ||
                    (!active_shadow_policies_.empty() && !streaming_shadows_) ||
                    (route_entry_ && route_entry_->internalRedirectPolicy().enabled());
-  if (buffering &&
+  if (buffering && !buffer_limit_exceeded_no_connection_ &&
       getLength(callbacks_->decodingBuffer()) + data.length() > retry_shadow_buffer_limit_) {
-    ENVOY_LOG(debug,
-              "The request payload has at least {} bytes data which exceeds buffer limit {}. Give "
-              "up on the retry/shadow.",
-              getLength(callbacks_->decodingBuffer()) + data.length(), retry_shadow_buffer_limit_);
-    cluster_->trafficStats()->retry_or_shadow_abandoned_.inc();
-    retry_state_.reset();
-    buffering = false;
-    active_shadow_policies_.clear();
-    request_buffer_overflowed_ = true;
-
-    // If we had to abandon buffering and there's no request in progress, abort the request and
-    // clean up. This happens if the initial upstream request failed, and we are currently waiting
-    // for a backoff timer before starting the next upstream attempt.
-    if (upstream_requests_.empty()) {
-      cleanup();
-      callbacks_->sendLocalReply(
-          Http::Code::InsufficientStorage, "exceeded request buffer limit while retrying upstream",
-          modify_headers_, absl::nullopt,
-          StreamInfo::ResponseCodeDetails::get().RequestPayloadExceededRetryBufferLimit);
-      return Http::FilterDataStatus::StopIterationNoBuffer;
+    if (wait_for_connect_) {
+      // we are still waiting for a connection, going over the buffer limit which will pause the
+      // read on the downstream connection
+      ENVOY_LOG(debug,
+                "Going over buffer limit {} due to wait on connect. buffer size: {}."
+                "This will pause read on downstream connection.",
+                retry_shadow_buffer_limit_,
+                getLength(callbacks_->decodingBuffer()) + data.length());
+      buffer_limit_exceeded_no_connection_ = true;
+      // ensure we stop on reading, the buffer limit might be greater than
+      // retry_shadow_buffer_limit_ due to an override per route of the retry_shadow_buffer limit
+      callbacks_->onDecoderFilterAboveWriteBufferHighWatermark();
```

Comment:
This does not necessarily hard stop data from entering the proxy, but it might stop this buffer from receiving more data. What is the limit on how much we're willing to buffer?

Reply:
That make perfect sense, I hadn't fully understand Alyssa comment.
 
I update the code to do what you suggest. 
Setting a hard limit of two times the upstream connection buffer limit. 
In you example, we can not set the limit to Y, the idea would be to reach the limit on the upstream buffer in order to:
- pause the read on the downstream connection.
- not block data anywhere in the proxy as it might have been the case before.

Does that make sense to you ?
two times the upstream buffer limit can probably be discussed or coming from a runtime value with an appropriate default